### PR TITLE
Enhance the checks in Encrypter constructor

### PR DESCRIPTION
**Pull request by ibrasho**
One issue I see with the length enforced checks is that if someone was using an invalid key (shorter or longer than they should) this would break BC for them. This applies to @GrahamCampbell RP https://github.com/laravel/framework/pull/9063 too

What do you think @taylorotwell ?

cc/ @ircmaxell @paragonie-scott

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
@ibrasho Doesn't Laravel 5.0 generate a 32 byte key for the AES-128-CBC cipher though? That's going to cause issues if we force people off it.

File: src/Illuminate/Encryption/Encrypter.php [hunk 1]
```diff
@@ -40,19 +50,20 @@ class Encrypter implements EncrypterContract
...
-            $this->cipher = $cipher;
-        } else {
-            throw new RuntimeException('The only supported ciphers are AES-128-CBC and AES-256-CBC.');
+        $this->cipher = $cipher;
+
+        $supportedKeyLength = $this->supportedCiphers[$cipher];
+        $keyLen = mb_strlen($key = (string) $key, '8bit');
+
+        if ($keyLen != $supportedKeyLength) {
+            throw new RuntimeException('Unsupported key length for '.$cipher.'. Use a '.$supportedKeyLength.' bytes key.');
```
Comment (GrahamCampbell) [inline src/Illuminate/Encryption/Encrypter.php]: Please use double quotes at keep the variables inside the string.

**Comment by ibrasho** [comment]
I know. In OpenSSL this get truncated, I'm check the mcrypt source code to see how it deals with keys of invalid length.

File: src/Illuminate/Encryption/composer.json [hunk 2]
```diff
@@ -16,6 +16,7 @@
     "require": {
         "php": ">=5.5.9",
         "ext-openssl": "*",
+        "ext-mbstring": "*",
```
Comment (GrahamCampbell) [inline src/Illuminate/Encryption/composer.json]: This change is unrelated, and not needed as we already require this extension because we require illuminate/support, which requires it.

**Comment by ibrasho** [comment]
Tests are failing do to invalid keys being used in tests.

**Comment by GrahamCampbell** [comment]
> Tests are failing do to invalid keys being used in tests.

lol

**Comment by GrahamCampbell** [comment]
Wait for Taylor to respond I think. Due to his timezone, that might be a few hours yet. :)

**Comment by GrahamCampbell** [comment]
This logic is overcomplicate. I'll send a new PR, and add more tests.

**Comment by franzliedke** [comment]
Huh? It's just two ifs.

**Comment by GrahamCampbell** [comment]
These changes are also wrong. We MUST allow 32 byte keys for the 128 cipher, otherwise everyone's encrypted data will be broken by our mac checks differing.

File: src/Illuminate/Encryption/composer.json (hunk 2 above)
Comment (crynobone) [inline src/Illuminate/Encryption/composer.json]: I would prefer to disagree on this, we should refer this as we refer `"php": ">=5.5.9"`, why do we include it here when illuminate/support already include the same line? Clarity, someone who probably looking for an encryption package found out about this package and decide to try it out because packagist only show it require php 5.5 & openssl extension.

Comment (GrahamCampbell) [inline src/Illuminate/Encryption/composer.json]: k, i'll change it

**Comment by ibrasho** [comment]
Why did we stop waiting for @taylorotwell to respond?

We still have the issue of people who are using keys with the wrong lengths. Both PRs will break BC for them.

We don't have a way to warn people about these issues, do we?

**Comment by GrahamCampbell** [comment]
> Why did we stop waiting for @taylorotwell to respond?

Taylor has responded, just not on GitHub. Just talked to him. :)

**Comment by ibrasho** [comment]
I would throw an `E_WARNING` error, but Laravel convert these to `ErrorException`s, right?

**Comment by ibrasho** [comment]
> We still have the issue of people who are using keys with the wrong lengths. Both PRs will break BC for them.

This point still needs a decision. Are we going to break BC for people who have chosen a bad key?

**Comment by GrahamCampbell** [comment]
> Are we going to break BC for people who have chosen a bad key?

No. Because we encouraged that bad default by mistake...

**Comment by ibrasho** [comment]
What about people who used keys or more than 16 bytes?

**Comment by paragonie-scott** [comment]
> No. Because we encouraged that bad default by mistake...

I hadn't anticipated this. But @ircmaxell mentioned something in another comment: Using HKDF to split the encryption and authentication keys.

HKDF(32 byte input) can give a 16 byte output. (It's not *recommended* to do this in reverse, however, because if you need a 32 byte key you should be using 32 bytes of entropy!)

This still won't be backwards compatible with data encrypted without HKDF, but it won't break 256-bit keys for AES-128.

**Comment by ircmaxell** [comment]
> These changes are also wrong. We MUST allow 32 byte keys for the 128 cipher, otherwise everyone's encrypted data will be broken by our mac checks differing.

You're confusing things. With mcrypt, key length determined which variant of AES you got. So Rijndael-128 with a 32 byte key was actually AES-256. That's because the number in Rijndael is block-size, while the number in AES is key-size (it's always a 128 bit block).

So if you were using a 32 byte key, you were using AES-256. So using that same 32 byte key should work transparently if you use AES-256-CBC...

Additionally, it would absolutely be worth introducing a version identifier into the cipher output so that you can identify and not have to worry about things like this anymore when you need to change it again in the future (and you will).

**Comment by ibrasho** [comment]
I think one piece of information that I forgot to mention is `AES ~= Rijndeal`. Some difference exists in the AES version of the algorithm.

This works:
```php
    $iv = mcrypt_create_iv(mcrypt_get_iv_size(MCRYPT_RIJNDAEL_128, MCRYPT_MODE_CBC), MCRYPT_DEV_URANDOM);
    $key = str_repeat('a', 32);
    $payload = 'payload';

    $e = mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $key, $payload, MCRYPT_MODE_CBC, $iv);

    // AES-256: This prints 'payload'
    echo openssl_decrypt($e, 'AES-256-CBC', $key, OPENSSL_RAW_DATA | OPENSSL_NO_PADDING, $iv);

    // AES-128: This doesn't work
    echo openssl_decrypt($e, 'AES-128-CBC', $key, OPENSSL_RAW_DATA | OPENSSL_NO_PADDING, $iv);
```

> @ircmaxell Additionally, it would absolutely be worth introducing a version identifier into the cipher output so that you can identify and not have to worry about things like this anymore when you need to change it again in the future (and you will).

Adding the cipher used to the final payload should be fine, right? Just like the IV currently.

**Comment by paragonie-scott** [comment]
> Adding the cipher used to the final payload should be fine, right?

I believe so. It probably doesn't even need to be MAC'd, esp. if the MAC algorithm can change in the future.

**Comment by taylorotwell** [comment]
Somewhere in all of this we've lost the ability to decrypt data that was encrypted by 5.0. That must be fixed or release will have to be delayed. I know it was working when we first made these changes.

**Comment by taylorotwell** [comment]
Check out the tests I worked on here: https://github.com/laravel/framework/commit/6cacd3e3cb43f1e672c992eaf8b4fadef36d9ac6

I added an McryptEncrypter which is the 5.0 branch version of the encrypter. This allows us to test the 5.0 implementation against the new OpenSSL implementation in 5.1. As you can see in the test, I encrypted the data using the MCRYPT_RIJNDAEL_128 of the 5.0 mcrypt encrypter, but needed to specify the AES-256-CBC cipher when using the new encrypter for it to pass the tests. The AES-128-CBC will fail... I find this weird since we used 128 in the 5.0 encrypter...

Explanation?

**Comment by taylorotwell** [comment]
Sounds like Anthony's comment a few posts up explains why I needed to use AES-256-CBC.

**Comment by GrahamCampbell** [comment]
https://github.com/laravel/framework/pull/9109 should fix everything remaining. :)

**Comment by ibrasho** [comment]
There seems to be no equivalent to `RIJNDEAL_192` or `RIJNDEAL_256` in OpenSSL. Which is why I suggested that we have two implementations of the encrypter.

Good job @GrahamCampbell :+1:

**Comment by GrahamCampbell** [comment]
The plan is to only have one implementation still. The other one is only there so we can verify the two work the same.

**Comment by ircmaxell** [comment]
@ibrasho using Rijndael 192 or 256 is not a good idea in the first place. It's not AES. And Rijndael 256 is actually susceptible to related key attacks. So no go.

> Adding the cipher used to the final payload should be fine, right? Just like the IV currently.

No. You want an algorithm identifier that specifies everything. Serialization, mac, how you use your keys, etc. Everything.

Take this scenario. You ship 5.1, everything's great. Come 6 months from now, a bug (security or not) is found in how you serialize requiring you to change your format. Nothing about the cipher will tell you that. Only the algorithm you deal with.

So creating an internal algorithm identifier(suggest 3 "magic bytes" (to reduce changes of matching random data) and 1 byte for the algorithm. Something like `"\x0F\x0F\x0F\x01"` where the "01" is the algorithm identifier for the first version of the algorithm. The next would be "02", etc...

**Comment by ibrasho** [comment]
> @ibrasho using Rijndael 192 or 256 is not a good idea in the first place. It's not AES. And Rijndael 256 is actually susceptible to related key attacks. So no go.

I'm worried about backward-compatibility. I surely won't advise they be the default, but I'm worried we might make someone's live harder than it should be.

**Comment by ibrasho** [comment]
> Take this scenario. You ship 5.1, everything's great. Come 6 months from now, a bug (security or not) is found in how you serialize requiring you to change your format. Nothing about the cipher will tell you that. Only the algorithm you deal with.

> So creating an internal algorithm identifier(suggest 3 "magic bytes" (to reduce changes of matching random data) and 1 byte for the algorithm. Something like "\x0F\x0F\x0F\x01" where the "01" is the algorithm identifier for the first version of the algorithm. The next would be "02", etc...

This is up to @taylorotwell to decide.

**Comment by taylorotwell** [comment]
I think the way we have it is fine for now. It doesn't have any backwards compatibility issues and moving forward we have the default of the OpenSSL AES implementation instead of the old Mcrypt RIJNDAEL implementation.

**Comment by ircmaxell** [comment]
> I'm worried about backward-compatibility. I surely won't advise they be the default, but I'm worried we might make someone's live harder than it needs to be.

Strongly disagree. If someone's using Rijndael-256, it's a bug. And a pretty significant one at that. It should be changed. That could be a documentation issue, but it should be changed.

As far as the algorithm identifier, I would definitely suggest doing that, as it will allow you to change over time in a safe backwards-compatible manner. That way you can improve security and robustness over time without having to worry about breaking anything.
